src/dataset.js

```diff
@@ -205,6 +208,22 @@ function Dataset(bigQuery, id) {
 
   this.bigQuery = bigQuery;
 
+  /**
+   * @name Dataset#metadata
+   * @type {object}
+   */
+  this.metadata = metadata || {};
+
+  /**
+   * @name Dataset#location
+   * @type {string}
+   */
+  Object.defineProperty(this, 'location', {
```

Comment:
This pattern is unusual to me. metadata could be considered a protected property that is only supposed to be a reflection of the remote object. Can we just localalize the "location" property of a user-provided options object, and then embed it later as part of whatever requests need it?

Reply:
Well, the issue I was facing was that if the location comes back through any `apiResponse` we should be using it, so in some cases the user might not even provide a location.

Reply:
Ah, gotcha. And if we didn't include it in follow up responses, it would error? If that's the case, it makes total sense to me! Could you just add a little inline comment to explain that?

Reply:
Unfortunately I won't be able to find the time to do that today. :(

Reply:
Ok, I've added a developer comment explaining this. PTAL